### DISCUSSION TITLE
Voters but for object capability decision

### DISCUSSION THREAD

**Discussion by mmarton** [symfony/symfony#52641] [Q&A]
Hi!

sorry for the meaningless title.
theoretical question:

we use Voter classes to determine if a **user is authorized to do some action (on a resource)**.
But what should I use to determine if an **action can be done on a resource at all**? Is it still Voters? Validation? Other?

for example I have a Program entity that can be archived only if it's closed and all invoices are in order and some others stuff checks out. It needs multiple services so I can't just create a function canBeArchived() inside the entity. It doesn't feel right to put this logic in a voter neither. so? WDYT?

**Comment by mbabker** [comment]
If you were using the Workflow component, this would be the type of thing you could plug in to block a transition - https://symfony.com/doc/current/workflow.html#blocking-transitions

There isn't really an out-of-the-box framework concept that fits into this type of app workflow beyond that.  You can do it with the validation component, you can do it with events, you can do it with a `canArchive()` method directly on the entity, you can use a custom service that anything that deals with that archive transition calls; there's no one right answer necessarily.
